docs/usecases/testing.md

```diff
@@ -72,3 +72,364 @@ Test aspects are powerful tools for modifying behavior of individual tests or ev
+  )
+}
+```
+There is also an easy way to test object's data for certain assertions with `hasField` which accepts besides a name, a mapping function from object to its tested property and `Assertion` object which will validate this property. Here our test checks if a person has at least 18 years and is not from USA. What is nice about those tests is that, test reporters will tell you exactly which assertion was broken. Let's say we would change `isGreaterThanEqualTo(18)` to `isGreaterThanEqualTo(40)` which will fail. Printout
+on console will be a nice detailed text what exactly went wrong:
+
+```bash
+[info]       User(Jonny,26,Address(Denmark,Copenhagen)) did not satisfy (hasField("age", _.age, isGreaterThanEqualTo(45)) && hasField("country", _.country, not(equalTo(USA))))
+[info]       26 did not satisfy isGreaterThanEqualTo(45)
+```
+
+
+Having this all in mind probably the most common and also most readable way of structuring tests is to pass
+a for-comprehension to `testM` function and yield a call to `assert` function.
+
+```scala
+testM("Semaphore should expose available number of permits") {
+  for {
+    s         <- Semaphore.make(1L)
+    permits   <- s.available
+  } yield assert(permits, equalTo(1L))
+}
+```
+
+
+_Running tests_
+
+When all of our test are constructed we need to have a way to actually execute them. Your first stop is the `zio.test.DefaultRunnableSpec` which accepts a single suite that will be executed. A single suite might seem to be limiting but as it was already said suit can hold any number of other suites. You may structure your tests like this:
+
+```scala
+val suite1 = suite("suite1") {
+  testM("s1.t1") {...},
+  testM("s1.t2") {...}
+}
+val suite2 = suite("suite2") {
+  testM("s2.t1") {...},
+  testM("s2.t2") {...},
+  testM("s2.t3") {...}
+}
+val suite3 = suite("suite3") {
+  testM("s3.t1") {...}
+}
+
+object AllSuites extends DefaultRunnableSpec(suite("All tests")(suite1, suite2, suite3))
+```
+
+`DefaultRunnableSpec` is very similar in its logic of operations to `zio.App`. Instead of providing one `ZIO` application
+at the end of the world we provide a suite that can be a tree of other suites and tests. Another resemblance is that `DefaultRunnableSpec` provides an Environment. Here it is an instance of `MockEnvironment` which helps us with controling our systems infrastructure. More info on using mocks can be found in sections bellow. 
+Just like with `zio.App` where at the very end an instance of `ZIO[R,E,A]` is expected where `R` can be at maximum of type `Environment` in `DefaultRunnableSpec` `R` cannot be more than `MockEnvironment`. So just like in normal application if our
+`R` is composed of some other modules we need to provide them first before test can be executed. How can we provide our dependencies?
+Here again the design of `zio-test` shines. Since our tests are ordinary values we can jest transform them with call to `mapTest`.
```

Comment:
```suggestion
Here again the design of `zio-test` shines. Since our tests are ordinary values we can just transform them with a call to `mapTest`.
```